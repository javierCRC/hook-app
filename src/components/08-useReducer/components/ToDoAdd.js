import React from 'react'
import { useForm } from '../../../hooks/useForm';
import { f_MessageProcess } from '../../../helpers';

export const ToDoAdd = ({ pF_handleAddTodo }) => {

    // custom hook for capture the value of each input of any form.
    const { stateValues , f_handleInputChange , f_handleCleanForm } = useForm({
        txtDescription:''
    });

    // destructuring the description value of stateValue for be use to individual form
    const { txtDescription } = stateValues;
    
    //console.log(txtDescription);

    const f_handleSubmitAddTodo = (e) => {
        e.preventDefault();

        if(txtDescription.trim().length <= 1){
           f_MessageProcess("Error empty field","Can't be empty fields","error");
           return;
        }

        const c_NewTodo = {
           id : new Date().getTime(),
           desc: txtDescription,
           done: false
       };

       
       pF_handleAddTodo ( c_NewTodo ); // sending the action to the reducer 

       f_handleCleanForm(); // for clean the form values before use the custom hook "useForm"

   }

    return (
        <>
                 <h4> Add ToDo <i className="far fa-list-alt"></i> </h4>
                                <hr />

                                <form    onSubmit={ f_handleSubmitAddTodo }
                                        className="bg-light my-3 p-3 pt-lg-0 border rounded">
                                   <h5 className="font-weight-bold text-center my-4 text-success"> Registry <i className="fas fa-calendar-day"></i></h5>     
                                
                                    <div className="form-group form-row">
                                        <label  className="col-12 col-form-label">Description:</label>

                                        <div className="input-group col-12 col-md-10">
                                            <div className="input-group-prepend">
                                                <span className="input-group-text text-primary"><i className="fas fa-id-card"></i></span>
                                            </div>
                                            <input type="text" 
                                                   name='txtDescription'
                                                   className="form-control"
                                                   autoComplete='off' 
                                                   placeholder="Whrite a new To Do" 
                                                   value={ txtDescription }
                                                   onChange={ f_handleInputChange }
                                                   />
                                        </div>
                                    </div>

                                    <div className="form-group form-row mb-0">
                                        <div className="col-auto">
                                            <button type="submit" 
                                                    className="btn btn-primary btn-block"
                                            > Save <i className="fas fa-save"></i>
                                            </button>
                                        </div>
                                    </div>

                                </form>
        
            
        </>
    )
}
