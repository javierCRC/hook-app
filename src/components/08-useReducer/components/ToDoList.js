import React from 'react'
import { ToDoListItem } from '../../08-useReducer/components/ToDoListItem';

export const ToDoList = ( { pTodo = [] , pF_handleToggleTask, pF_f_handleClickDelete } ) => {
    return (
        <ul>
            {
                                        
                   pTodo.map( ( vTodo , i) => ( 
                    <ToDoListItem 
                                   key = { pTodo.id } 
                                   pToDo2={ vTodo } 
                                   pIndex={ i } 
                                   pF_handleToggleTask2={ pF_handleToggleTask } 
                                   pF_f_handleClickDelete2={ pF_f_handleClickDelete }
                    />
                ))
            }   
        </ul>
    ) 
}
