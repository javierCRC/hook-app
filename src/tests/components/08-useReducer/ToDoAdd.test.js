import { shallow } from "enzyme";

import { arrToDos } from "../../fixtures/demoToDos";
import { ToDoAdd } from '../../../components/08-useReducer/components/ToDoAdd';
import '@testing-library/jest-dom'; 


describe('Test of <ToDoAdd />', () => {
    
     
    const vF_handleAddTodo = jest.fn(); 
    
 
     const cWrapper = shallow(<ToDoAdd pF_handleAddTodo = { vF_handleAddTodo  }          
                               />
                                     );
    
    test('1.Must do render the component ToDoAdd.', () => {
        
        // 1 Arrange *******************************************************************************
            
        //2. Act *******************************************************************************
             //console.log(cWrapper.html());

        //3. Assert *******************************************************************************
              expect(cWrapper).toMatchSnapshot();

    });

    test('2.Must dont be call the function pF_handleAddTodo.', () => {
        
        // 1 Arrange *******************************************************************************
             let vFormSubmit;
             

        //2. Act *******************************************************************************
             vFormSubmit = cWrapper.find('form').prop('onSubmit');
                //console.log('function submit: '+ vFormSubmit );  
            vFormSubmit({ preventDefault(){} }) // another form to call a function  

        //3. Assert *******************************************************************************
             expect(vF_handleAddTodo).toHaveBeenCalledTimes(0);
            
    });

    test('3.Must be correctly call the function pF_handleAddTodo simulating change and submit.', () => {
        
        // 1 Arrange *******************************************************************************
            let vFormChange1;
        
            let vFormSubmit;
                
            const c_SimulateEvent  = {
                target: {
                    name: 'txtDescription',
                    value: 'Learn React with JS in 2022'
                }
            };


            const c_NewTodo = {
                id : expect.any(Number),
                desc: 'Learn React with JS in 2022',
                done: false
            };

            let vTxtDescription;
             

        //2. Act *******************************************************************************
                  vFormChange1 = cWrapper.find('input').simulate('change',c_SimulateEvent ); // form 1 to simulate a event
                     //console.log('function change 1: '+ vFormChange1 );
                 
                  vFormSubmit = cWrapper.find('form').prop('onSubmit');
                     //console.log('function submit: '+ vFormSubmit );  
                  vFormSubmit({ preventDefault(){} }) // another form to call a function 
                  
                  vTxtDescription = cWrapper.find('input').prop('value');
                
        //3. Assert *******************************************************************************
              expect(vF_handleAddTodo).toHaveBeenCalledTimes(1);
              expect(vF_handleAddTodo).toHaveBeenCalledWith(c_NewTodo);
              expect(vTxtDescription).toBe(''); // this for test the functionality of f_handleCleanForm
            
    });

    
})
