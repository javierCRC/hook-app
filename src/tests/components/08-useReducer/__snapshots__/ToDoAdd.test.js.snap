// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas de <ToDoAdd /> 1.Must do render the component ToDoAdd. 1`] = `
<Fragment>
  <h4>
     Add ToDo 
    <i
      className="far fa-list-alt"
    />
     
  </h4>
  <hr />
  <form
    className="bg-light my-3 p-3 pt-lg-0 border rounded"
    onSubmit={[Function]}
  >
    <h5
      className="font-weight-bold text-center my-4 text-success"
    >
       Registry 
      <i
        className="fas fa-calendar-day"
      />
    </h5>
    <div
      className="form-group form-row"
    >
      <label
        className="col-12 col-form-label"
      >
        Description:
      </label>
      <div
        className="input-group col-12 col-md-10"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text text-primary"
          >
            <i
              className="fas fa-id-card"
            />
          </span>
        </div>
        <input
          autoComplete="off"
          className="form-control"
          name="txtDescription"
          onChange={[Function]}
          placeholder="Whrite a new To Do"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group form-row mb-0"
    >
      <div
        className="col-auto"
      >
        <button
          className="btn btn-primary btn-block"
          type="submit"
        >
           Save 
          <i
            className="fas fa-save"
          />
        </button>
      </div>
    </div>
  </form>
</Fragment>
`;

exports[`Test of <ToDoAdd /> 1.Must do render the component ToDoAdd. 1`] = `
<Fragment>
  <h4>
     Add ToDo 
    <i
      className="far fa-list-alt"
    />
     
  </h4>
  <hr />
  <form
    className="bg-light my-3 p-3 pt-lg-0 border rounded"
    onSubmit={[Function]}
  >
    <h5
      className="font-weight-bold text-center my-4 text-success"
    >
       Registry 
      <i
        className="fas fa-calendar-day"
      />
    </h5>
    <div
      className="form-group form-row"
    >
      <label
        className="col-12 col-form-label"
      >
        Description:
      </label>
      <div
        className="input-group col-12 col-md-10"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text text-primary"
          >
            <i
              className="fas fa-id-card"
            />
          </span>
        </div>
        <input
          autoComplete="off"
          className="form-control"
          name="txtDescription"
          onChange={[Function]}
          placeholder="Whrite a new To Do"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group form-row mb-0"
    >
      <div
        className="col-auto"
      >
        <button
          className="btn btn-primary btn-block"
          type="submit"
        >
           Save 
          <i
            className="fas fa-save"
          />
        </button>
      </div>
    </div>
  </form>
</Fragment>
`;
